microhapulator/Snakefile

```diff
@@ -0,0 +1,369 @@
+                data = b64encode(fh.read()).decode("ascii")
+                r2readlenplots.append(data)
+            with open(f"analysis/{sample}/{sample}-merged-read-lengths.png", "rb") as fh:
+                data = b64encode(fh.read()).decode("ascii")
+                mergedreadlenplots.append(data)
+            with open(f"analysis/{sample}/{sample}-interlocus-balance.png", "rb") as fh:
+                data = b64encode(fh.read()).decode("ascii")
+                locbalanceplots.append(data)
+            with open(f"analysis/{sample}/{sample}-heterozygote-balance.png", "rb") as fh:
+                data = b64encode(fh.read()).decode("ascii")
+                hetbalanceplots.append(data)
+        templatefile = resource_filename("microhapulator", "data/template.html")
+        with open(templatefile, "r") as infh, open("report.html", "w") as outfh:
+            template = Template(infh.read())
+            output = template.render(
+                date=datetime.now().replace(microsecond=0).isoformat(),
+                mhpl8rversion=microhapulator.__version__,
+                samples=config["samples"],
+                summary=summary,
+                r1readlenplots=r1readlenplots,
+                r2readlenplots=r2readlenplots,
+                mergedreadlenplots=mergedreadlenplots,
+                locbalanceplots=locbalanceplots,
+                hetbalanceplots=hetbalanceplots,
+                static=5,
+                dynamic=0.02,
+                zip=zip,
+            )
+            print(output, file=outfh, end="")
+
+
+rule summary:
+    input:
+        expand("analysis/{sample}/flash.log", sample=config["samples"]),
+        expand("analysis/{sample}/{sample}-mapped-reads.txt", sample=config["samples"]),
+        expand(
+            "analysis/{sample}/fullrefr/{sample}-fullrefr-mapped-reads.txt",
+            sample=config["samples"],
+        ),
+        expand("analysis/{sample}/{sample}-typing-rate.tsv", sample=config["samples"]),
+        expand("analysis/{sample}/{sample}-interlocus-balance-chisq.txt", sample=config["samples"]),
+        expand(
+            "analysis/{sample}/{sample}-heterozygote-balance-pttest.txt",
+            sample=config["samples"],
+        ),
+    output:
+        tsv="analysis/summary.tsv",
+    run:
+        columns = (
+            "Sample",
+            "TotalReads",
+            "Merged",
+            "MergeRate",
+            "Mapped",
+            "MappingRate",
+            "MappedFullRefr",
+            "MappingRateFullRefr",
+            "Typed",
+            "TypingRate",
+            "InterlocChiSq",
+            "HetTstat",
+        )
+        data = {colname: [] for colname in columns}
+        for sample in sorted(config["samples"]):
+            print(f"[Compiling summary] Sample={sample}", file=sys.stderr)
+            totalreads, mergedreads, *_ = parse_flash_summary(f"analysis/{sample}/flash.log")
```

Comment:
What does the `*_` do? Put's the return values of `parse_flash_summary()` into a list?

Reply:
Syntactic sugar: if a function returns many values, you can grab the first N values in named variables and then cram anything else into a list or tuple. The `*` indicates "the rest of the values, if any" and the `_` is a nondescript variable name.

Reply:
It's surprisingly hard to get clear documentation for this syntax. It's a `*` prefix operator (as opposed to the infix operator for multiplication). Probably the best keywords to search for are "variable packing/unpacking" and "variadic".

[This page](https://treyhunner.com/2018/10/asterisks-in-python-what-they-are-and-how-to-use-them/) provides a pretty nice explainer: it's worth reading the whole thing, but the section on "Asterisks in tuple unpacking" discusses this particular feature.